sources/org.osbuild.curl

```diff
@@ -123,19 +127,27 @@ def fetch(url, checksum, directory):
             if return_code == 0:
                 break
         else:
-            raise RuntimeError(f"curl: error downloading {url}: error code {return_code}")
-
-        if not verify_file(f"{tmpdir}/{checksum}", checksum):
-            raise RuntimeError(f"checksum mismatch: {checksum} {url}")
-
-        # The checksum has been verified, move the file into place. in case we race
-        # another download of the same file, we simply ignore the error as their
-        # contents are guaranteed to be  the same.
-        try:
-            os.rename(f"{tmpdir}/{checksum}", f"{directory}/{checksum}")
-        except FileExistsError:
-            pass
-
+            error = f"curl: error downloading, last return code {return_code}"
+
+        if not error and not verify_file(f"{tmpdir}/{checksum}", checksum):
+            error = f"checksum mismatch: {checksum}"
+
+        if not error:
+            # The checksum has been verified, move the file into place. In case
+            # we race another download of the same file, we simply ignore the
+            # error as their contents are guaranteed to be  the same.
+            try:
+                os.rename(f"{tmpdir}/{checksum}", f"{directory}/{checksum}")
+            except FileExistsError:
+                error = "redundant downloading of the same file"
+
+        if error:
+            download_status["error"] = error
```

Comment:
It would be neat to have this be more structured, rather than just a string. The curl error code at least, maybe other data we have, if any? Or we could translate the curl error codes to our own enum if that's more appropriate.

Reply:
Thanks for the review! Yeah you are right! the return code should also be a member of `DownloadResult` Here I just recycled the old exception message. I will integrate that in the next iteration.